enhancements/authentication/allowing-uri-scheme-in-oidc-sub-claims.md

```diff
@@ -0,0 +1,167 @@
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Make the oauth-server capable of setting up identities for OIDC identity
+providers that use the URI scheme in the `sub` claim of their ID tokens.
+
+## Motivation
+
+Per the [OIDC specification](https://openid.net/specs/openid-connect-core-1_0-final.html#IDToken)
+of the ID token and its more thorough details described in
+[RFC7519 - JSON Web Token (JWT)](https://tools.ietf.org/html/rfc7519#section-4.1.2),
+the `sub` claim may be represented as both a string or a URI. However, URI by
+definition may contain characters such as `/` and `:` that are hard to digest
+for the `identity.oauth.openshift.io` API, which prevents integration with OIDCs
+that make use of this URI scheme. However, an identity provider that follows all
+specifications should still be usable as such for OpenShift to successfully retrieve
+identities from.
+
+### Goals
+
+1. Make it possible for users of the aforementioned identity providers to log in
+   to OpenShift successfully.
+
+### Non-Goals
+
+None.
+
+## Problem Summary
+
+With OpenShift 4.x, the oauth-server no longer allows [configuring the claim
+of ID tokens](https://github.com/openshift/api/blob/670ac3fc997c2f1d19b8c29ef04f70d6e3d4a59e/osin/v1/types.go#L341)
+retrieved from the OIDC provider to use as the unique identifier
+for the given user. This identifier is later used to construct the name of an
+`Identity` object that describes the mapping between the identity provider the
+user is a member of, and the user.
+
+The name of the `Identity` object is constructed by using the following pattern:
+
+```text
+<identity provider name from config>:<unique user identifier from the identity provider>
+```
+
+This naming is further validated so that neither the username, nor the identity provider
+name may contain the `:` character, and, according to Kubernetes naming schemes,
+neither can also contain the `/` sign, yet both of these commonly appear as a
+part of URI/URL schemes.
+
+## Proposal
+
+Due to the API limitations mentioned in [Problem Summary](#problem-summary), a
+generic `sub` claim cannot be safely used in order to create an `Identity` object.
+It is apparent that these will need to be encoded in a certain way. Simply encoding
+a `sub` claim won't do as that would break authentication for any OIDC user
+that had its `Identity` created prior to the version containing this fix.
+
+Proposal: if a `sub` claim of an ID token contains either `:` or `/`, url-encode
+the value of the claim and create the new identity object like so:
+
+```yaml
+apiVersion: user.openshift.io/v1
+kind: Identity
+metadata:
+  name: <identity provider name>:url:<url-encoded sub claim value>
```

Comment:
What is the resource path of such an object with encoding?

Reply:
`/apis/user.openshift.io/v1/foo:url:https%3A%2F%2Fheise.de%0A` is not valid I assume, is it?

Reply:
that'd be something like: thisbeuri%3A%3A32423423%2F23432
I should probably change that to base64: dGhpc2JldXJpOjozMjQyMzQyMy8yMzQzMg==

WDYT?

Reply:
b64 certainly makes it easier to reason about. Have you tried to understand exactly and write down here how url encoding will end up in the URL path? Am not saying that it does not work. Am just curious to see an example here.

Reply:
Hm, you're right, %s really won't make for a good, valid name. The more I think about it the more I want to use url-safe base64 instead. I'll change that.